### ISSUE TITLE
[BUG] Cannot access a disposed object.  Object name: 'SslStream' when when downloading files asynchronously using a blobClient in batch through a proxy

### ISSUE THREAD

**Issue by hikarunoryoma** [Azure/azure-sdk-for-net#18483]
Creating an issue as a response to @pakrym: https://github.com/Azure/azure-sdk-for-net/issues/17527#issuecomment-774161963 

**Describe the bug**
When reusing a BlobClient to download multiple files through a proxy(10 files that are around 18Mb each) in batch I hit an error that I can not access a disposed SslStream.

**Expected behavior**
I should be able to reuse a blobClient to asynchronously download blobs in batch.

**Actual behavior (include Exception or Stack Trace)**
The following error: 
```
System.ObjectDisposedException: Cannot access a disposed object.  Object name: 'SslStream'.     at
System.Net.ConnectStream.EndRead(IAsyncResult asyncResult) at
System.Threading.Tasks.TaskFactory`1.FromAsyncTrimPromise`1.Complete(TInstance thisRef, Func`3 endMethod, IAsyncResult asyncResult, Boolean requiresSynchronization)  --- End of stack trace from previous location where exception was thrown ---     at 
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()     at 
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)     at 
System.Net.Http.HttpClientHandler.WebExceptionWrapperStream.<ReadAsync>d__4.MoveNext()  --- End of stack trace from previous location where exception was thrown ---     at 
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()     at 
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)     at 
Azure.Core.Pipeline.ReadTimeoutStream.<ReadAsync>d__5.MoveNext()  --- End of stack trace from previous location where exception was thrown ---     at 
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()     at 
Azure.Core.Pipeline.RetriableStream.RetriableStreamImpl.<ReadAsync>d__12.MoveNext()  --- End of stack trace from previous location where exception was thrown ---     at 
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()     at 
Azure.Core.Pipeline.RetriableStream.RetriableStreamImpl.<RetryAsync>d__13.MoveNext()  --- End of stack trace from previous location where exception was thrown ---     at 
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()     at 
Azure.Core.Pipeline.RetriableStream.RetriableStreamImpl.<ReadAsync>d__12.MoveNext()  --- End of stack trace from previous location where exception was thrown ---
```

**To Reproduce**
```csharp
private async Task Main(string[] args)
{
	var uri = new UriBuilder("myUrl")
	{
		Query =
			"mySas"
	}.Uri;
	var options = new BlobClientOptions
	{
		Transport = new HttpClientTransport(new HttpClientHandler
		{
			Proxy = new WebProxy
			{
				Address = new Uri("myProxy")
			}
		})
	};
	var azureBlobClient = new BlobServiceClient(uri, options);
	var containerClient = azureBlobClient.GetBlobContainerClient("myContainer");
	var azureBlobNames = new List<string>();
	// GetBlobAsync itself is not awaitable, but enumerating items is an awaitable action
	foreach (var blob in containerClient.GetBlobs(
			prefix:
			@"myPrefix")
		)
		if (blob.Name.EndsWith(".parquet"))
			azureBlobNames.Add(blob.Name);
	using (var semaphore = new SemaphoreSlim(10))
	{
		var fileDownloadTasks = azureBlobNames.Select(azureBlobName => Task.Run(async () =>
		{
			// claim one
			await semaphore.WaitAsync(CancellationToken.None);
			try
			{
				await DownloadFileAsync(
					containerClient,
					azureBlobName
				);
			}
			finally
			{
				// ensure we always release
				semaphore.Release();
			}
		}, CancellationToken.None)).ToArray();
		await Task.WhenAll(fileDownloadTasks);
	}
}
private async Task DownloadFileAsync(
	BlobContainerClient containerClient,
	string blobName
)
{
	// strip out the relative path from blobName since DownloadToAsync does not generate directories
	var destinationFilePath = $@"{Path.GetFileName(blobName)}";
	var blobClient = containerClient.GetBlobClient(blobName);
	using (var stream = new FileStream(destinationFilePath, FileMode.OpenOrCreate))
	{
	       // fails here
		var download = await blobClient.DownloadAsync(CancellationToken.None);
		using (var downloadInfo = download.Value)
		{
			await downloadInfo.Content.CopyToAsync(stream);
		}
	}
}
```

We can get around this by not using a Semaphore and doing downloads sequentially or instantiating a new blobClient for each blob file, but neither solution is desirable.  The example below is instantiating a new blobClient for each blob file which works, but opens up sockets that aren't immediatly reclaimed or reused, hence our hesitance to use this since our production machines might serve multiple requests (with multiple blobs) which doesn't scale

```csharp
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;
using Azure.Core.Pipeline;
using Azure.Storage.Blobs;
namespace ConsoleApp1
{
	public class Program
	{
		public static async Task Main(string[] args)
		{
			var azureBlobNames = new List<string>();
			using (var handler = new HttpClientHandler
			{
				Proxy = new WebProxy
				{
					Address = new Uri("myProxy")
				}
			})
			{
				var options = new BlobClientOptions
				{
					Transport = new HttpClientTransport(handler)
				};
				var uri = new UriBuilder("myUrl")
				{
					Query =
						"mySas"
				}.Uri;
				var azureBlobClient = new BlobServiceClient(uri, options);
				var containerClient = azureBlobClient.GetBlobContainerClient("myContainer");
				// GetBlobAsync itself is not awaitable, but enumerating items is an awaitable action
				foreach (var blob in containerClient.GetBlobs(
						prefix:
						@"myPrefix")
					)
					if (blob.Name.EndsWith(".parquet"))
						azureBlobNames.Add(blob.Name);
			}
			using (var semaphore = new SemaphoreSlim(int.Parse(args[0])))
			{
				var fileDownloadTasks = azureBlobNames.Select(azureBlobName => Task.Run(async () =>
				{
					// claim one
					await semaphore.WaitAsync(CancellationToken.None);
					try
					{
						await DownloadFileAsync(
							azureBlobName
						);
					}
					finally
					{
						// ensure we always release
						semaphore.Release();
					}
				}, CancellationToken.None)).ToArray();
				await Task.WhenAll(fileDownloadTasks);
			}
		}
		private static async Task DownloadFileAsync(
			string blobName
		)
		{
			// strip out the relative path from blobName since DownloadToAsync does not generate directories
			var destinationFilePath = $@"{Path.GetFileName(blobName)}";
			using (var handler = new HttpClientHandler
			{
				Proxy = new WebProxy
				{
					Address = new Uri("myProxy")
				}
			})
			{
				var options = new BlobClientOptions
				{
					Transport = new HttpClientTransport(handler)
				};
				var uri = new UriBuilder("myUrl")
				{
					Query =
						"mySas"
				}.Uri;
				var azureBlobClient = new BlobServiceClient(uri, options);
				var containerClient = azureBlobClient.GetBlobContainerClient("container");
				var blobClient = containerClient.GetBlobClient(blobName);
				using (var stream = new FileStream(destinationFilePath, FileMode.OpenOrCreate))
				{
					var download = await blobClient.DownloadAsync(CancellationToken.None);
					using (var downloadInfo = download.Value)
					{
						await downloadInfo.Content.CopyToAsync(stream);
					}
				}
			}
		}
	}
```

**Environment:**
 - Name and version of the Library package used: Azure.Storage.Blobs 12.7.0
 - Hosting platform or OS and .NET runtime version: Windows 10, .NET Framework 4.7.2
 - IDE and version : VS2019 version 16.5.0

**Comment by hikarunoryoma** [comment]
My belief is that the changes made to `HttpWebRequestTransport` should fix all requests (not only upload), but the changelog suggest otherwise: https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/CHANGELOG.md#added-2.

I would like confirmation if my hypothesis is true, but if not this might need further investigation.

Thanks!

**Comment by jsquire** [comment]
Thank you for your feedback.  Tagging and routing to the team member best able to assist.

**Comment by pakrym** [comment]
The `HttpWebRequestTransport` fix won't help here because you are not using the `HttpWebRequestTransport`, instead, you are using the `HttpClientTransport`.

**Comment by hikarunoryoma** [comment]
Ah, yea that makes sense. Let me know if you need anything else on my end to help investigate!

**Comment by pakrym** [comment]
We are shipping an updated Azure.Core in a few days that includes a change in the timeout handling. I'd like you to try it and see if it helps.

**Comment by hikarunoryoma** [comment]
Great, I'll comment back here with the results and will be eagerly awaiting the update!

**Comment by malhotrar** [comment]
So @hikarunoryoma and I were able to resolve this by setting MaxConnectionsPerServer = 10+ as part of the HttpClientTransport - is there a specific reason why setting that works? We will note that the data source we are reaching out to is in another continent and thus the downloads are slower than our other clients using that same exact code above.

**Comment by hikarunoryoma** [comment]
We also found that setting the `MaxConnectionsPerServer = 10 ` seems to work for us. 

Setting the BlobClientOptions like so worked, but curious as to why the difference in behavior.
```
var options = new BlobClientOptions
			{
				Transport = new HttpClientTransport(new HttpClientHandler
				{
					Proxy = new WebProxy
					{
						Address = new Uri("http://10.20.6.20:8080")
					},
					MaxConnectionsPerServer = 10
				})
			};
```

**Comment by pakrym** [comment]
Sorry for the very long response. I've somehow missed the notification.

> Is there a specific reason why setting that works? 

I suspect that this exception is caused by the request timing out and by increasing the parallel request count you allowed pending requests to go through quicker and not time out. 

But we should've been suppressing these exception in Azure.Core if they are caused by timeouts we control. I'll look into it.

**Comment by malhotrar** [comment]
@pakrym  - not a problem - like we mentioned above, we unblocked ourselves by setting MaxConnectionsPerServer and have yet to encounter another issue since then (It was great you wrote this blog post: https://devblogs.microsoft.com/azure-sdk/net-framework-connection-pool-limits/)

**Comment by AlexGhiondea** [comment]
@pakrym did you end up figuring out what's going on here?
